lib/Dialect/ATen/ATenLoweringPass.cpp

```diff
@@ -0,0 +1,972 @@
+                              Value i = ivs[0];
+                              Value j = ivs[1];
+                              Value k = ivs[2];
+                              Value l = ivs[3];
+                              iRes(i, j, k, l) =
+                                  iLHS(i, j, k, l) + iRHS(i, j, k, l);
+                            });
+    }
+    // Return the newly allocated buffer.
+    rewriter.replaceOp(op, {result});
+    return success();
+  }
+};
+
+
+// Replace the given operation with a call to the given function.
+// The function is assumed to accept memrefs and scalar types and return
+// Memrefs. Here the result types are converted back to the result types of op,
+// but operands are NOT converted.  This allows non-standard mappings from
+// operand types to function types.
+LogicalResult
+rewriteWithVoidFunctionCallExplicit(Operation *op,
+                                    ArrayRef<Value> callops,
+                                    ArrayRef<Value> operands,
+                                    ConversionPatternRewriter &rewriter,
+                                    std::string functionName) {
+
+    auto loc = op->getLoc();
+    edsc::ScopedContext scope(rewriter, loc);
+
+    // The original operation types.
+    SmallVector<Type, 8> opTys;
+    // Shape erased versions of the original operation types.
+    SmallVector<Type, 8> erasedOpTys;
+    for (auto &o: callops) {
+      Type t = o.getType();
+      opTys.push_back(t);
+      if (t.isa<MemRefType>())
+        erasedOpTys.push_back(getShapeErasedMemRefType(t.cast<MemRefType>()));
+      else
+        erasedOpTys.push_back(t);
+    }
+
+    std::vector<Value> newOps = callops;
+    SmallVector<Value, 8> newResults;
+
+    // Result types of the original operation, converted to memrefs.
+    SmallVector<Type, 8> retTys;
+    // Erased version of the return type.  This is the return types of the
+    // generated function call.
+    SmallVector<Type, 8> erasedRetTys;
+    for (const auto &o: op->getResults()) {
+      Type t = o.getType();
+      if (t.isa<TensorType>()) {
+        TensorType tensorResultTy = t.cast<TensorType>();
+        MemRefType memRefResultTy =
+          mlir::MemRefType::get(tensorResultTy.getShape(),
+                                tensorResultTy.getElementType(), {}, 0);
+        MemRefType erasedMemRefResultTy = getShapeErasedMemRefType(memRefResultTy);
+        retTys.push_back(memRefResultTy);
+        //        erasedRetTys.push_back(erasedMemRefResultTy);
+        // assume memRefResultTy has known shape, so we don't need any
+        // dynamic dimensions for the alloc.
+        assert(memRefResultTy.hasStaticShape());
+        Value allocVal = rewriter.create<AllocOp>(op->getLoc(),
+                                                  memRefResultTy);
+        Value castVal = MemRefTypeCast(rewriter, allocVal);
+        newOps.push_back(castVal);
+        newResults.push_back(allocVal);
+    } else {
+        assert(false);
```

Comment:
Is there a reason why we are asserting here (and above) vs returning a failing result?